Feature/web scraping

mterada1228 opened the pull request:
- 修正内容
    - WebScrapingのために以下のGemを追加
      - `nokogiri`
      - `httparty`
    - Horseモデルをスクレイピングするスクリプトを作成
      - 取得サイト例: https://db.netkeiba.com/horse/2016105089

mterada1228 committed:
horseモデルのscraping scriptを追加

ershn commented on Gemfile.lock:
```diff
@@ -283,10 +342,13 @@ DEPENDENCIES
+  nokogiri (~> 1.10, >= 1.10.10)
   pg
   pry-rails
   puma (~> 4.1)
```
こちらのファイルは最新バージョンではないかと思います

mterada1228 replied:
`bundle update`を実施し、こちらを最新化いたしました。

ershn replied:
OK

ershn commented on app/script/web_scraping/horse_scraper.rb:
```diff
@@ -0,0 +1,21 @@
+require 'HTTParty'
```
こちらは不要かと思います。

mterada1228 replied:
`app/services/scraper/horse_scraper_service.rb`に移動しましたが、  
先頭行の `require 'HTTParty'`を削除致しました。

ershn replied:
OK

ershn commented on app/script/web_scraping/horse_scraper.rb:
```diff
@@ -0,0 +1,21 @@
+  doc = Nokogiri::HTML(html)
+
+  {
+    id: url.gsub(/\d+$/).first,
```
`gsub`は文字列内の一部を変えたい時に使うメソッドですが、データを抽出したい時は `/pattern/ =~ url` のようにした方が適切だと思います。

Ruby 2.2.7 のものですが、以下を一旦ご確認ください。
https://ruby-doc.org/core-2.2.7/Regexp.html

mterada1228 replied:
app/services/scraper/horse_scraper_service.rbに移動しましたが、
以下のように修正いたしました。

```ruby
id: /\d+$/.match(url)[0]
```

ershn replied:
OK

ershn commented on app/script/web_scraping/horse_scraper.rb:
```diff
@@ -0,0 +1,21 @@
+
+  {
+    id: url.gsub(/\d+$/).first,
+    name: doc.css('div.horse_title > h1').children[0].text.gsub(/[[:space:]]/, '')
```
Chrome の Developer tools で css セレクターのコピー機能を活かせば、情報の抽出が楽になると思います。

mterada1228 replied:
ご連携ありがとうございます！活用いたします。

ershn replied:
OK

ershn commented on app/script/web_scraping/horse_scraper.rb:
```diff
@@ -0,0 +1,21 @@
+  }
+end
+
+def create(data)
```
`data`という変数名は何を指すのかはわからないので、より具体的な名前にしたほうが保守性があがります。

mterada1228 replied:
app/services/scraper/horse_scraper_service.rbに移動しましたが、
`data` => `attribute` に変更致しました。

ershn replied:
OK

ershn commented on app/script/web_scraping/horse_scraper.rb:
```diff
@@ -0,0 +1,21 @@
+ActiveRecord::Base.logger = Logger.new($stdout)
+
+html = HTTParty.get(ARGV[0])
+create(parse(html))
```
コマンドラインで呼ばれるのを想定するスクリプトだと思いますが、サービスクラスの形にしたほうがのちほど別の形での再利用が楽になります。

mterada1228 replied:
app/services/scraper/horse_scraper_service.rbに移動しました。

ershn replied:
OK

mterada1228 committed:
指摘対応
- bundle update を実行し、Gemfile.lock を最新化
- horse_scraper.rb を services/scraper 配下に移動
- database.yml を修正し、test環境のDBをdevelop環境と同様のmysqlに変更

mterada1228 commented:
先週はレビューありがとうございました。
修正致しましたのでご確認お願い致します。

mterada1228 committed:
Feature/web scraping
- HorseRaceResult は複合主キーを止め、ナチュラルキーを主キーとして追加
- HorseRaceResultに完走できなかった理由を入れる`reason_of_exclusion`カラムを追加
- Scraping スクリプトの `html` => `response` に変数名を変更
- RaceResult, HorseRaceResult のスクレイピングスクリプトを追加

mterada1228 commented on app/services/scraper/horse_scraper_service.rb:
```diff
@@ -0,0 +1,36 @@
+    end
+
+    def call
+      html = HTTParty.get(url)
```
`html`という名前というよりは`response` や `request` の方がふさわしい

ershn replied:
OK

mterada1228 committed:
指摘対応
- html => response に変数名を修正

ershn commented on app/services/scraper/horse_scraper_service.rb:
```diff
@@ -0,0 +1,35 @@
+      }
+    end
+
+    def create(attribute)
```
アトリビュートは複数含まれてるので、`attributes`の方が正しいです。

mterada1228 replied:
`app/services/scraper/*` で使用していた変数 `attribute` は `attributes` に変更しました。

ershn replied:
OK

ershn commented on app/services/scraper/race_result_scraper_service.rb:
```diff
@@ -0,0 +1,106 @@
+    def initialize(url:)
+      @url = url
+      @wait_time = 3
+      @timeout = 4
```
こちらは動的に変わることがないので、定数にしておいたほうが適切です。

mterada1228 replied:
`@wait_time`の使用箇所はなかったので削除しました。  
`@time_out` は `TIMEOUT` の定数に変更しました。

ershn replied:
OK

ershn reviewed (commented):
差分で不要なものが出ないように依存となってる PR をこちらの PR にマージし、diff 対象をその PR にしてもらえればと思います。

ershn commented on app/services/scraper/horse_race_result_scraper_service.rb:
```diff
@@ -0,0 +1,140 @@
+        {
+          horse_id: %r{horse/(\d+)}.match(url)[1],
+          race_result_id: %r{race/(.+)/\z}
+            .match(race_result.children[9].children[0].attributes['href'].value)[1],
```
DOM にアクセスするための低レベルなインターフェイスを使うと、編集コストが上がって、可読性も下がるので、ここは css セレクターか xpath を使えれば楽になると思います。

mterada1228 replied:
css セレクターを使用して Element を取得するように変更しました。

ershn replied:
OK

mterada1228 committed:
Merge branch 'develop' into feature/web-scraping

mterada1228 committed:
refs Feature/web scraping 指摘対応
- HORSE_RACE_RESULTテーブルにupdateができるように、ナチュラルキーのIDを追加
- attribute => attributes に変数名を変更
- app/services/scraper/race_result_scraper_service.rbのインスタンス変数を定数に変更
- DOM へのアクセサを css selector に変更

mterada1228 commented on db/schema.rb:
```diff
@@ -10,9 +10,9 @@
+ActiveRecord::Schema.define(version: 2020_12_17_050441) do
 
-  create_table "horse_race_results", primary_key: ["horse_id", "race_result_id"], options: "ENGINE=InnoDB DEFAULT CHARSET=utf8", force: :cascade do |t|
+  create_table "horse_race_results", id: :bigint, unsigned: true, options: "ENGINE=InnoDB DEFAULT CHARSET=utf8", force: :cascade do |t|
```
修正前は `horse_id` と `race_result_id` の複合主キーとなっていました。  
このままだと `ActiveRecord` のメソッドが使用できないため、  
ナチュラルキーのIDを主キーとするように変更致しました。

ershn replied:
ActiveRecord はデフォルトで複合キーをサポートしてないので、composite_primary_keys の gem を入れて、主キーをモデルで指定すれば、エラーが解消されるかもしれません。

https://github.com/composite-primary-keys/composite_primary_keys

このモデルでは複合キーで正しいと思います。

mterada1228 replied:
- Gemfileにconposite_primary_keysを追加しました。

- migationファイルを使用して、1. idカラムの削除、2. horse_id, primary_key を複合プライマリキーに変更しました。  

  - migration files:  
    - db/migrate/20210120013903_delete_id_and_add_composite_prymary_key_to_horse_id_and_race_id_to_horse_race_result.rb
    - db/migrate/20210121042512_add_foreign_key_horse_id_on_horse_race_result.rb

ershn replied:
OK

mterada1228 reviewed (commented):
@ershn レビューありがとうございました。
修正を行いましたので改めてレビューをお願い致します。

mterada1228 committed:
refs feature/web-scraping Raceモデルのスクレイピングを追加

mterada1228 commented on app/models/race.rb:
```diff
@@ -4,8 +4,31 @@
   has_many :race_horses
   has_many :horses, through: :race_horses
+  has_many :race_regulations
+  has_many :race_prizes
```
レコードを横持ちから縦持ちに変更するために以下のように変更しました。

(旧)  
`Race` に `regulation1, regulation2, regulation3...` と `prize1, prize2, prize3...` のカラムを持つ  
(新)
`RaceRegulation`, `RacePrize` モデルを追加して `race_id` で紐付ける

ershn replied:
了解です。いい修正だと思います。

mterada1228 commented on app/models/race.rb:
```diff
@@ -4,8 +4,31 @@
+  enum course: {
+    sapporo: 1, hakodate: 2, fukushima: 3, niigata: 4, tokyo: 5,
+    nakayama: 6, cyuukyou: 7, kyouto: 8, hanshin: 9, kokura: 10 }
+  enum grade: { g1: 0, g2: 1, g3: 2, Listed: 3, op: 4, three_wins: 5, two_wins: 6, one_win: 7 }
```
入力値が限定されるカラムに `enum` を追加しました。

ershn replied:
OK

mterada1228 commented on app/models/race.rb:
```diff
@@ -4,8 +4,31 @@
+  SIDE_TRANSLATIONS = {
+    '内' => Race.sides[:inner_course],
+    '外' => Race.sides[:external_course]
+  }.freeze
```
スクレイピングした値を、enum値に変換するためのハッシュです。

ershn replied:
OK

mterada1228 commented on db/schema.rb:
```diff
@@ -71,28 +88,21 @@
   create_table "races", id: :bigint, default: nil, options: "ENGINE=InnoDB DEFAULT CHARSET=utf8", force: :cascade do |t|
     t.datetime "start", null: false
-    t.string "course", null: false
+    t.integer "course", null: false
```
`enum` に変換したので `string` => `integer` に変更しております。

ershn replied:
OK

mterada1228 commented on db/schema.rb:
```diff
@@ -71,28 +88,21 @@
     t.integer "round", null: false
     t.string "name", null: false
-    t.string "grade"
+    t.integer "grade"
```
`enum` に変換したので `string` => `integer` に変更しております。

ershn replied:
OK

mterada1228 reviewed (commented):
@ershn Race モデルのスクレイピングを追加しましたのでレビューをお願い致します。  
(取得元ページ例: https://race.netkeiba.com/race/shutuba.html?race_id=202006050711)

ershn commented on Gemfile.lock:
```diff
@@ -290,32 +281,31 @@ DEPENDENCIES
+  tzinfo-data (~> 1.2.7)
+  web-console (~> 3.3.0)
   webdrivers (~> 4.4, >= 4.4.1)
   webpacker (~> 4.0)
```
`bundle update`で色々更新されたという話かと思いますが、それで何が更新されるのかわからなくて、ものが意図せず壊れてしまうこともあるので、`bundle update`はやめて、必要な更新だけをした方が安全です。

mterada1228 replied:
承知いたしました。
以降そのように対応致します。

ershn replied:
OK

ershn commented on app/models/race.rb:
```diff
@@ -4,8 +4,31 @@
+    '内' => Race.sides[:inner_course],
+    '外' => Race.sides[:external_course]
+  }.freeze
 end
```
レコードの間で重複するデータがあるように思いますので、重複がないように見直した方が簡単になると思います。

mterada1228 replied:
以下の方針で見直ししました。
1. `race_result`の重複するカラムを削除
     - `name`, `course_id`, `course_length`, `date`, `course_type`, `prize`
2. `race_result` に `race_id` を参照するようにカラムを追加
3.  `race_result`を参照していた`horse_race_result` の参照は`race`に変更

ershn replied:
いいモデリングだと思います。

ershn commented on app/models/race.rb:
```diff
@@ -4,8 +4,31 @@
+  enum course: {
+    sapporo: 1, hakodate: 2, fukushima: 3, niigata: 4, tokyo: 5,
+    nakayama: 6, cyuukyou: 7, kyouto: 8, hanshin: 9, kokura: 10 }
+  enum grade: { g1: 0, g2: 1, g3: 2, Listed: 3, op: 4, three_wins: 5, two_wins: 6, one_win: 7 }
```
`Listed`だけに大文字が入ってます。

mterada1228 replied:
`Listed` => `listed` に修正致しました。

ershn replied:
OK

mterada1228 committed:
指摘対応

ershn commented on app/models/horse.rb:
```diff
@@ -4,6 +4,8 @@
-  has_many :races, through: :race_horses
+  # TODO
+  ## race_horse は廃止。horse_race_result は horse_race に修正して統一する
+  has_one :race, through: :horse_race_results
```
has_many の方が正しいでは？

mterada1228 replied:
`has_one` から `has_many` に relation を変更しました.

ershn replied:
OK

ershn commented on app/services/scraper/horse_race_result_scraper_service.rb:
```diff
@@ -0,0 +1,139 @@
+
+      race_results.map do |race_result|
+        attributes = {}
+        OPERATOR.each_key do |column_name|
```
`OPERATOR.each do |column, value|`でもいいかと思います。

mterada1228 replied:
ご指摘の通りに修正を実施いたしました。

ershn replied:
OK

ershn commented on app/services/scraper/horse_race_result_scraper_service.rb:
```diff
@@ -0,0 +1,139 @@
+      race_results.map do |race_result|
+        attributes = {}
+        OPERATOR.each_key do |column_name|
+          attributes[column_name] = OPERATOR[column_name].call elements race_result
```
ここのメソッド呼び出しで括弧を使った方が読みやすいと思います。

---

`elements(race_result)`はそれほど軽くなさそうなので、ループの外でやったほうがいいと思います。

mterada1228 replied:
- `HorseRaceResultScraperService`
  - ループ外に`elements(rase_result)`呼び出しを記述しました
- それ以外の `** ScraperService`
  - `|=` 演算子を使用して、処理は一度しか呼ばれないように修正いたしました

ershn replied:
OK

ershn reviewed (changes requested):
コメントを確認と対応してから、マージしていただいて問題ないです。ありがとうございます。

mterada1228 committed:
feature/web-scraping 指摘対応
	-  HorseRaceResultモデルはhorse_id, race_idの複合キーに変更する
		- composite_primary_keys gem の追加
		- DB定義を変更
	- HorseとRaceのrelationはhas_manyに変更
	- HorseRaceResultScraperServiceのOPARATORループは each do |k, v| にリファクタリング
	- **ScraperServiceのelementsメソッド呼び出しは一度しか呼び出されないように修正

ershn commented on app/services/scraper/race_scraper_service.rb:
```diff
@@ -0,0 +1,96 @@
+        day_number: doc.css('#page > div.RaceColumn01 > div > div.RaceMainColumn > div.RaceList_NameBox > div.RaceList_Item02 > div.RaceData02 > span:nth-child(3)'),
+        regulations_and_prizes: doc.css('#page > div.RaceColumn01 > div > div.RaceMainColumn > div.RaceList_NameBox > div.RaceList_Item02 > div.RaceData02')
+      }
+    end
```
`@var ||=`のキャッシュはできる場合は避けるべきで、クラスはできるだけステートレスな作りにする方が保守性が上がります。

ステートが増えれば増えるほど、バグが発生してしまう可能性があがるので。

---

`url`はコンストラクタ内に定義されて更新されないので、問題ないですが、`doc`は動的に変わるものなので、`elements`がその変数に依存することとして明示的に引数から受け取った方がわかりやすく保守しやすいです。

mterada1228 replied:
- `@elements ||=`呼び出しは削除しました。 
- `doc`はインスタンス変数ではなく、ローカル変数に変更しました。

ershn replied:
OK

ershn commented on app/services/scraper/race_result_scraper_service.rb:
```diff
@@ -0,0 +1,109 @@
+
+    # rubocop:disable Layout/LineLength
+    def elements
+      @elements |= {
```
ここは`||=`ですね。

mterada1228 replied:
https://github.com/mterada1228/keiba_expectation_app/pull/10#discussion_r562357831 で `@elements ||=` 呼び出しは削除しました。

ershn replied:
OK

mterada1228 committed:
refs feature/web-scraping 指摘対応
- **ScraperServiceをステートレスにするため、docをインスタンス変数ではなくローカル変数に変更
- @elements ||= のキャッシュはバグを誘発する可能性があるので削除
